### DISCUSSION TITLE
One or more?

### DISCUSSION THREAD

**Discussion by pawamoy** [DanCardin/cappa#185] [Q&A] [answered]
I can't seem to find whether it's possible to require one or more arguments to be consumed. There's only `num_args=-1` for zero or more. I tried using `num_args=-1, required=True` but this didn't change anything.

**Comment by DanCardin** [comment]
There's always

```python
class Foo:
    foo_required: An[int, Arg(value_name='foo')]
    foo_rest: An[list[int], Arg(value_name='foo')]

    @property
    def foo(self):
        return [self.foo_required, *self.foo_rest]
```

I was going to suggest this (but it doesn't work correctly atm, for a reason).

```python
class Foo:
    foo: An[list[int], Arg(num_args=1, group=Group()), Arg(group=Group())]
```

And there's also `parse` as an option for disallowing empty lists manually.

---

As far as a "native" feature, with `required`...i'll have to attempt to make that modification to the parser and see what breaks, if anything.

**Reply by pawamoy** [reply]
> foo_required, foo_rest

Not very elegant :sweat_smile: 

> Arg(), Arg()

It feels a bit weird to specify arg two times. Seems complex to support in a robust manner?

Anyway I see you have opened #186. I don't know if `num_args=-1` + `required=True` is the best solution, but it would work for me :slightly_smiling_face:

**Reply by DanCardin** [reply]
> Not very elegant 😅

heh, yea. but i **think** off the top of my head in other parsers you'd either need 2 args and the same kind of thing, or a custom action, or a post-hoc validation. So this option is roughly equivalent to what you'd have to do elsewhere anyways, but at least it keeps the logic self contained to the type.

At least, assuming combining required=True with num_args=-1 isn't a configuration click/argparse support, i'd have to check.

---

the 2-arg syntax was originally for mutually exclusive args (which is why here you also have to explicitly state a non-exclusive `Group()`), but the net effect is that it mostly already works (i mean it doesnt, but it could probably fairly straightforwardly). So i'm not so concerned about complexity there so much as it's verbose and perhaps not the clearest.

---

if it works, and in particular if that's the behavior you get from argparse (if argparse allows you to do that), then i'm happy with that as an api, since it requires explicit opt in.

**Reply by pawamoy** [reply]
> in other parsers you'd either need 2 args

Not that you should mimic `argparse`, but it supports `nargs="+"` :smile: It actually provides constants for this:

```pycon
>>> import argparse
>>> argparse.ZERO_OR_MORE
'*'
>>> argparse.ONE_OR_MORE
'+'
```

Could do something similar and hide `-1` behind `cappa.ZERO_OR_MORE` and `-2` (probably?) behind `cappa.ONE_OR_MORE`.

**Reply by DanCardin** [reply]
blegh. so long as argparse supports it, i feel decently about it. but i think (personally) `num_args=-1, required=True` is a preferable api. I very much wanted to keep the type strictly at integers, and ideally have no magical sentinel values (which -1 unfortunately is)

Although tbh, i'm not sure off the top of my head why `num_args=0` isn't the obvious magical sentinel value here...

**Reply by pawamoy** [reply]
> I very much wanted to keep the type strictly at integers, and ideally have no magical sentinel values (which -1 unfortunately is)

yep, totally understand that feeling

I also wondered if 0 wouldn't be a better match, but then how would you say "do not consume any additional arg", when the type is a list for example?

**Reply by DanCardin** [reply]
mm i'm not sure what you mean there. but i think it's exclusively used right now for options like flags or `--help`. and because those could all be handled by looking at the action; in actual reality, it's probably only relevant if you set it to zero and defined a custom `action=`.

**Reply by pawamoy** [reply]
hmmm yeah not sure what I meant either :sweat_smile: I do use `num_args=0` for args with custom actions: `--version` (print version and exit) and `--debug-info` (print debug info and exit).

**Reply by DanCardin** [reply]
okay so actually, it turns out the argparse backend **exclusively** uses `'+'` because (afaict) there's no point in `'*'` as long as you provide a `default` (which we do).

So the current behavior using the argparse backend with:

```
@dataclass
class Args:
    a: list[str]
```

already **is** a one-or-more behavior, and it's just a behavior difference bug vs the native parser.

And in fact a sequence-type argument without a default is **already** inferred as `required=True`, we just weren't doing anything with the `required` flag unless it was an option before.

That seem right? seems right to me i think.

**Reply by pawamoy** [reply]
That seems right indeed :thinking: So it means that, once the inferred `required` is correctly used:

- a `list[str]` arg with no default will require one or more args
- to make it optional again, we'll have to pass `required=False`
- in both cases we don't need to pass `num_args=-1` anymore (not sure I got this one right)

Sounds much cleaner :smile:

**Reply by DanCardin** [reply]
to make it optional, you'd have to provide a default, which would then infer required=False. there's a preexisting validation that validates you have a default if you explicitly specify `required=False`.

This is (iirc) because not all types have an obvious default cappa cna choose, and if you're explicitly saying it's not required, that implies **someone** needs to choose a default.

but yes, no need to specify num_args.

`foo: list[str]` == `Arg(num_args=-1, required=True)`
`foo: list[str] = field(default_factory=list)` == `Arg(num_args=-1, required=False, default=Default(ValueFrom(list)))`
`foo: Anntoated[list[str], Arg(required=False)]` == raise ValueError`

**Reply by pawamoy** [reply]
OK, perfect, thanks!

**Reply by DanCardin** [reply, marked as answer]
https://github.com/DanCardin/cappa/pull/187
